Add an ability to select Perlin noise for better consistency

kabachuha opened the pull request:
Before that the frame was being showered with uniform noise leading to loss of detail and consistency over time in contrary of it's intended objective of improving detalization.

Now users can select more diverse Perlin noise

A sample:

![perlin](https://user-images.githubusercontent.com/14872007/209855032-7097eb5f-3514-447a-9d63-bde6842e8458.png)

which in my experience improved the overall coherence

If it works well, it can be perfectly set to be the default noise mode.

With Perlin noise:

https://user-images.githubusercontent.com/14872007/209855283-b352a2eb-0501-421a-8703-be6cbf7c5159.mp4

Without Perlin noise:

https://user-images.githubusercontent.com/14872007/209855301-64ea6952-b9e8-45a8-a142-5041fc40492a.mp4

--

Addresses #29 and #174, as well as, to my mind, a lot of other issues indirectly

kabachuha committed:
add 2D Perlin noise

kabachuha committed:
add noise params to args

kabachuha committed:
safer noise_type loading from settings

kabachuha committed:
fix run args

kabachuha committed:
typo

kabachuha committed:
fixup args and tensor shapes to make it work

hithereai commented:
Frame #X with perlin noise:
![20221228211917_00113](https://user-images.githubusercontent.com/121192995/209862066-0ab1ffd4-dc93-4bad-922e-e16b69aaeedc.png)

Same frame with the old persistent noise:
![20221228211814_00113](https://user-images.githubusercontent.com/121192995/209862093-dc751fe8-4395-4897-9977-e1486c40cf1a.png)

Thoughts? And explanations about the new perlin-params would be great. 
Also - how does (and if) it play with noise_schedule?

kabachuha commented:
noise_schedule is the multiplier of how much total noise is applied on the image

roughly, `image = image + (white_noise * perlin_noise) * noise_schedule`

To make the 'noise cells' not too small, it is generated on lower resolutions (8x8 by default) and then upscaled to the image size

Octaves can be thought as how many perlin noise interations are layered one upon another, persistence as the layering modifier

![perlin](https://user-images.githubusercontent.com/14872007/209863537-75e53e33-b81b-4344-a73b-1670a0847df0.png)

^ This example was created from 8x8 with 8 octaves and 0.5 persistence

kabachuha commented:
As you see, Perlin noise tends to create 'organic shapes' whereas uniform noise makes more 'straight' forms

hithereai commented:
I'm getting the following error after a few single frames when changing perlin-octaves from 4 to 8:

```
File "D:\D-SD\GitGUITests\stable-diffusion-webui/extensions/deforum-for-automatic1111-webui/scripts\deforum_helpers\noise.py", line 14, in <lambda>
    dot = lambda grad, shift: (torch.stack((grid[:shape[0],:shape[1],0] + shift[0], grid[:shape[0],:shape[1], 1] + shift[1]  ), dim = -1) * grad[:shape[0], :shape[1]]).sum(dim = -1)
RuntimeError: The size of tensor a (768) must match the size of tensor b (512) at non-singleton dimension 1
```

It works fine with the default 4 octaves though. Is that expected? Cuz in your example you used 8 octaves

kabachuha commented:
Interesting... I'll check that out

kabachuha commented:
Breaks for me too at 8, though 7 works fine. Probably, it's just the algorithm limitation.

I advice you to keep this number lower as the more octaves it gets, the more spread-out it becomes which spoils the intended effect

hithereai commented:
Thanks for elaborating. 

Maybe we should keep hard-limits on params that are known to break things completely at certain value ranges?

There's another one in the anti-blur thaht breaks above 5 I think, and prolly plenty more to find. 🔢

kabachuha commented:
I can make sliders for these limited values

kabachuha committed:
make perlin octaves a slider

kabachuha committed:
add tooltips for perlin noise settings

kabachuha committed:
set perlin noise by default

it's just too good and addresses many beginner questions, especially related to frame coherence, detail loss and picture flattening as Perlin noise's organic nature counters all of them

kabachuha committed:
typo fix and tooltip wording improvement

kabachuha committed:
add note about Perlin noise to the header

kabachuha committed:
add more useful links to the header

kabachuha committed:
link perlin noise note to wikipedia

kabachuha committed:
fixup saved settings loading

kabachuha commented:
Now everything works fine
